### PR TITLE
Added Provisioner and unittest

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: This PR adds the Provisioner for the new runtime. This is an initial provisioner so it's rather straight forward. Later work can be done to improve the provisioning algorithm. 
*Testing*: Added new unittest which uses the CPU backend and CPUDeviceManager.
*Documentation*:
Related to #2045

### CODE REVIEW

File: lib/Runtime/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+add_subdirectory(Provisioner)
```
Comment (bertmaher) [inline lib/Runtime/CMakeLists.txt]: trivial: end files with newlines please

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+#include "glow/Backends/BackendUtils.h"
+#include "glow/Backends/CompiledFunction.h"
+#include "glow/Graph/Graph.h"
+
+#include <future>
+#include <queue>
+
+using namespace glow;
+using namespace runtime;
+using DeviceID = unsigned int;
```
Comment (bertmaher) [inline lib/Runtime/Provisioner/Provisioner.cpp]: RuntimeTypes.h has `using DeviceIDTy = size_t;`, is this device ID the same or something different?  If different, can you make the type names more distinctive?

Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: It's actually not used anywhere, so I'm removing it.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    nextNode.push(newSet);
+  }
+  while (!nextNode.empty()) {
+    FunctionMapTy compiledFunctions;
+    std::vector<std::unique_ptr<CompiledFunction>> functions;
+    auto nodes = nextNode.front();
+    nextNode.pop();
+    // Add child nodes to the queue.
+    for (unsigned int i = 0; i < nodes[0]->children.size(); i++) {
```
[the remark this answers is not in the document]
Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: it does not, what matters is that the right groups of nodes are paired and enqueued together, (the idea being they are share constants.) The improved provisioner will use the DAGNodes logical deviceID to group functions.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    // Check if sufficient space on device. Currently requiring a buffer
+    // over the size of constants determined by NETWORK_PADDING_FACTOR.
+    auto availableMemory = currDevice->second->getAvailableMemory();
+    auto requiredMemory = NETWORK_PADDING_FACTOR *
+                          nodes[0]->runtimeBundle.getConstantWeightSize();
+    if (availableMemory < requiredMemory) {
+      return ResultCode::Failed;
+    }
+    // Load functions on device. Create a promise that is passed into the
```
[the remark this answers is not in the document]
Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: To enable flexibility. If the deviceManager blocks we lose the option. It's likely a later rendition of Provisioner will take advantage of this async nature. First pass is simple so it blocks and waits.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    currDevice->second->addNetwork(
+        &module, compiledFunctions,
+        [&addNetwork](const Module *, ResultCode result) {
+          addNetwork.set_value(result == ResultCode::Ready);
+        });
+    auto result = ready.get();
+    if (!result) {
+      return ResultCode::Failed;
+    }
+    currDevice++;
```
[the remark this answers is not in the document]
Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: Good catch it's supposed to be outside the while loop. Moving it. 
It's a little convoluted, I inserted some newlines to help break it out. Essentially we are walking all related function trees in lockstep.

**Pull request merged**